File: deepeval/tracing/trace_context.py
from typing import Optional, List, Dict, Any
from contextvars import ContextVar
from contextlib import contextmanager
from dataclasses import dataclass

from .tracing import trace_manager
from .context import current_trace_context, update_current_trace
from deepeval.prompt import Prompt
from deepeval.metrics import BaseMetric
from deepeval.test_case.llm_test_case import ToolCall


@dataclass
class LlmSpanContext:
    prompt: Optional[Prompt] = None
    metrics: Optional[List[BaseMetric]] = None
    metric_collection: Optional[str] = None
    expected_output: Optional[str] = None
    expected_tools: Optional[List[ToolCall]] = None
    context: Optional[List[str]] = None
    retrieval_context: Optional[List[str]] = None


current_llm_context: ContextVar[Optional[LlmSpanContext]] = ContextVar(
    "current_llm_context", default=LlmSpanContext()
)


@contextmanager
def trace(
    llm_span_context: Optional[LlmSpanContext] = None,
    name: Optional[str] = None,
    tags: Optional[List[str]] = None,
    metadata: Optional[Dict[str, Any]] = None,
    thread_id: Optional[str] = None,
    user_id: Optional[str] = None,
    input: Optional[Any] = None,
    output: Optional[Any] = None,
    retrieval_context: Optional[List[str]] = None,
    context: Optional[List[str]] = None,
    expected_output: Optional[str] = None,
    tools_called: Optional[List[ToolCall]] = None,
    expected_tools: Optional[List[ToolCall]] = None,
    metrics: Optional[List[BaseMetric]] = None,
    metric_collection: Optional[str] = None,
):
    current_trace = current_trace_context.get()

    if not current_trace:
        current_trace = trace_manager.start_new_trace()

    if metrics:
        current_trace.metrics = metrics

    if metric_collection:
        current_trace.metric_collection = metric_collection
    
    current_trace_context.set(current_trace)

    update_current_trace(
        name=name, 
        tags=tags, 
        metadata=metadata, 
        thread_id=thread_id, 
        user_id=user_id, 
        input=input, 
        output=output, 
        retrieval_context=retrieval_context,
        context=context,
        expected_output=expected_output,
        tools_called=tools_called,
        expected_tools=expected_tools,
    )


    if llm_span_context:
        current_llm_context.set(llm_span_context)
    try: 
        yield current_trace
    finally:
        current_llm_context.set(LlmSpanContext())

File: typescript/src/tracing/trace-context.ts
import { AsyncLocalStorage } from "async_hooks";

import {
  getCurrentTrace,
  setCurrentTrace,
  Trace,
  traceManager,
  updateCurrentTrace,
} from "./tracing";

import { Prompt } from "../prompt";
import { ToolCall } from "../test-case";
import type { BaseMetric } from "../metrics/base-metrics";
import { _setAmbientPayloadReader } from "./pending-context";

export type LlmSpanContext = {
  prompt?: Prompt;
  metrics?: BaseMetric[];
  metricCollection?: string;
  toolsMetricCollection?: string;
  toolsMetrics?: BaseMetric[];
  expectedOutput?: string;
  expectedTools?: ToolCall[];
  context?: string[];
  retrievalContext?: string[];
};

export type AgentSpanContext = {
  metrics?: BaseMetric[];
  metricCollection?: string;
  expectedOutput?: string;
  expectedTools?: ToolCall[];
  context?: string[];
  retrievalContext?: string[];
};

const llmSpanContextStore = new AsyncLocalStorage<LlmSpanContext>();
export function getLlmContext(): LlmSpanContext | undefined {
  return llmSpanContextStore.getStore();
}
export function setLlmContext(context: LlmSpanContext | null): void {
  llmSpanContextStore.enterWith(context ?? {});
}

const agentSpanContextStore = new AsyncLocalStorage<AgentSpanContext>();
export function getAgentContext(): AgentSpanContext | undefined {
  return agentSpanContextStore.getStore();
}
export function setAgentContext(context: AgentSpanContext | null): void {
  agentSpanContextStore.enterWith(context ?? {});
}

_setAmbientPayloadReader((kind) => {
  const drop = (payload: Record<string, unknown>) => {
    const out: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(payload)) {
      if (value !== undefined) out[key] = value;
    }
    return Object.keys(out).length > 0 ? out : undefined;
  };

  if (kind === "llm") {
    const ctx = getLlmContext();
    if (!ctx) return undefined;
    return drop({
      metrics: ctx.metrics,
      metricCollection: ctx.metricCollection,
      prompt: ctx.prompt,
      expectedOutput: ctx.expectedOutput,
      expectedTools: ctx.expectedTools,
      context: ctx.context,
      retrievalContext: ctx.retrievalContext,
    });
  }

  if (kind === "tool") {
    // `toolsMetricCollection`/`toolsMetrics` are declared on the LLM context but
    // target the tool spans in scope.
    const ctx = getLlmContext();
    if (!ctx) return undefined;
    return drop({
      metrics: ctx.toolsMetrics,
      metricCollection: ctx.toolsMetricCollection,
    });
  }

  if (kind === "agent") {
    const ctx = getAgentContext();
    if (!ctx) return undefined;
    return drop({
      metrics: ctx.metrics,
      metricCollection: ctx.metricCollection,
      expectedOutput: ctx.expectedOutput,
      expectedTools: ctx.expectedTools,
      context: ctx.context,
      retrievalContext: ctx.retrievalContext,
    });
  }

  return undefined;
});

export async function setTracingContext<T>(
  opts: Partial<Trace> & {
    llmSpanContext?: LlmSpanContext;
    agentSpanContext?: AgentSpanContext;
  },
  fn: () => Promise<T> | T,
): Promise<T> {
  let currentTrace = getCurrentTrace();
  if (!currentTrace) {
    currentTrace = traceManager.startNewTrace();
  }

  if (opts.metrics) {
    currentTrace.metrics = opts.metrics;
  }
  if (opts.metricCollection) {
    currentTrace.metricCollection = opts.metricCollection;
  }

  setCurrentTrace(currentTrace);
  updateCurrentTrace({
    name: opts.name,
    tags: opts.tags,
    metadata: opts.metadata,
    threadId: opts.threadId,
    userId: opts.userId,
    testCaseId: opts.testCaseId,
    turnId: opts.turnId,
    input: opts.input,
    output: opts.output,
    retrievalContext: opts.retrievalContext,
    context: opts.context,
    expectedOutput: opts.expectedOutput,
    toolsCalled: opts.toolsCalled,
    expectedTools: opts.expectedTools,
    metrics: opts.metrics,
    metricCollection: opts.metricCollection,
  });

  if (opts.llmSpanContext) {
    setLlmContext(opts.llmSpanContext);
  }
  if (opts.agentSpanContext) {
    setAgentContext(opts.agentSpanContext);
  }

  return await fn();
}
